addons/misra.py

```diff
@@ -1353,6 +1398,46 @@ def misra_7_3(self, rawTokens):
             if compiled.match(tok.str):
                 self.reportError(tok, 7, 3)
 
+    def misra_7_4(self, data):
+        # A string literal shall not be assigned to an object unless the object's type
+        # is constant.
+        def reportErrorIfVariableIsNotConst(variable, stringLiteral):
+            if variable.valueType:
+                if variable.valueType.constness != 1:
+                    self.reportError(stringLiteral, 7, 4)
+
+        for token in data.tokenlist:
+            if token.isString:
+                # Check normal variable assignment
+                variable = getAssignedVariableToken(token)
+                if variable:
+                    reportErrorIfVariableIsNotConst(variable, token)
+                
+                # Check use as return value
+                function = getFunctionUsingReturnValue(token)
+                if function:
+                    # "Primitive" test since there is no info available on return value type
+                    if not tokenFollowsSequence(function.tokenDef, ['const', 'char', '*']):
+                        self.reportError(token, 7, 4)
+
+            # Check use as function parameter
+            if isFunctionCall(token):
+                parametersUsed = getArguments(token)
+                functionDeclaration = getFunction(data.functions,
```

Comment:
Example code:

    void foo(const char *p);

    void bar() {
       foo("123");
    }

The dump output for the function call:

    <token id="0x5642001f5da0" file="1.c" linenr="5" column="3" str="foo" scope="0x5642001d1cc0" type="name" function="0x5642001e5a00" astParent="0x5642001f4da0"/>
    <token id="0x5642001f4da0" file="1.c" linenr="5" column="6" str="(" scope="0x5642001d1cc0" link="0x5642001f4420" astOperand1="0x5642001f5da0" astOperand2="0x5642001f5000" valueType-type="void" valueType-reference="None"/>
    <token id="0x5642001f5000" file="1.c" linenr="5" column="7" str="&quot;123&quot;" scope="0x5642001d1cc0" type="string" strlen="3" values="0x5642001d0770" astParent="0x5642001f4da0" valueType-type="char" valueType-pointer="1" valueType-constness="1" valueType-reference="None"/>
    <token id="0x5642001f4420" file="1.c" linenr="5" column="12" str=")" scope="0x5642001d1cc0" link="0x5642001f4da0"/>
    <token id="0x5642001f41c0" file="1.c" linenr="5" column="13" str=";" scope="0x5642001d1cc0"/>

The function `function="0x5642001e5a00"` is available in the dump so you should be able to fetch this directly.

Reply:
Yes it was my first strategy to use the function reference from the token, but I ran into a problem:

Example code:
```
void foo(const char *p);
void bar(char *p);

void main() {
   foo("123");
   bar("123");
}
```

The dump output for the function call then produces:
```
<token id="000001480F448D80" file="1.c" linenr="5" column="4" str="foo" scope="000001480F3D27C0" type="name" function="000001480D6D9610" astParent="000001480F448C00"/>
    ...
<token id="000001480F448D20" file="1.c" linenr="6" column="4" str="bar" scope="000001480F3D27C0" type="name" astParent="000001480F449440"/>
```

I am not sure why I do not get a function reference in the _bar_ case. If I change the signature of the bar function and add a _const_ it works well. So maybe some other part of the system detects that the signature is wrong (missing const) but this is what I want to test. To work around this problem I used the less elegant approach and added the _getFunction_ based on name and parameter list.

Reply:
> So maybe some other part of the system detects that the signature is wrong (missing const) but this is what I want to test.

yes.. we need to tweak Cppcheck for this. I believe it makes sense to keep the behavior for C++ code but we need to tweak the behavior for C code.

this should be fixed asap.

If you want to look at tweaking Cppcheck yourself feel free to do it. I guess the function that needs to be adjusted is `Scope::findFunction` in lib/symboldatabase.cpp.